R/utils.R

```diff
@@ -22,16 +22,28 @@ base_pkgs <- function() {
 #' @importFrom pkgcache ppm_snapshots
 get_ppm_snapshot_by_date <- function(date) {
   snaps <- pkgcache::ppm_snapshots()
-  res <- as.character(as.Date(head(
+  res <- as.character(as.Date(utils::head(
     snaps[as.Date(snaps$date) > as.Date(date), "date"],
     1
   )))
-  if (length(res) == 0) stop(sprintf("Cannot find PPM snapshot for date after %s.", as.character(date)))
+  if (length(res) == 0) {
+    warning(sprintf(
+      paste0(
+        "Cannot find PPM snapshot for date after %s.",
+        " Will use current CRAN instead."
+      ),
+      as.character(date)
+    ))
+    return(NA)
+  }
   res
 }
 
 #' @importFrom pkgcache ppm_repo_url
-parse_ppm_url <- function(snapshot) {
+parse_ppm_url <- function(snapshot = NA) {
+  if (is.na(snapshot)) {
+    return(pkgcache::default_cran_mirror())
```

Comment:
Fallback to current CRAN in case package published date is NA (this applies directly in case workaround is reverted)

Reply:
There are few bad things happening here:
- `get_ppm_snapshot_by_date` warns that CRAN will be used but this is actually handled by a separate function `parse_ppm_url`. There will be nothing wrong if the latter will be called inside but that's not the case.
- Can we return NULL instead of NA. I kind of dislike NA as it is ambigious.
- From what I recall, "/latest" PPM endpoint was suppose to handle all edge cases (including Bioc). Shall we use cran mirror instead? Right now we have two different fallbacks and I am confused
- I would prefer to handle PPM urls inside PPM related functionalities and fallback into cran mirror outside. Something like: `url <- parse_ppm_snapshot(...) %||% pkgcache::default_cran_mirror()`. IMHO that would be much cleaner.

Reply:
This is a fallback for very recent versions that are after the latest `ppm` snapshot.

That is, releases from the last few days may not have a snapshot on `ppm`, this is a rare edge case, but this was happening with a very fresh release.

Reply:
Refactored the code:
* Output is clear as it returns the URL for the snapshot
  * `parse_ppm_url` is called inside instead
* Handles fallback inside instead of repeated call when this function is called
* Prefer using a parameter as I was testing if the CRAN is really needed or `latest`
  * Couldn't replicate the edge case that made me use CRAN mirror
    * When it depends on very recent packages _(1 or 2 day old)_
    * Latest ppm seems synced with `pkgcache::meta_cache_list`
    * This can be tested by using packages on top of [this list](https://cran.r-project.org/web/packages/available_packages_by_date.html)

https://github.com/insightsengineering/verdepcheck/blob/0819cbdbce04b16be2fe60f921d8677508d241d1/R/utils.R#L27-L49